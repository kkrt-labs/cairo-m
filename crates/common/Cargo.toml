[package]
name = "cairo-m-common"
version = "0.1.0"
edition.workspace = true

[dependencies]
num-traits.workspace = true
paste = "1.0"
serde.workspace = true
serde_json.workspace = true
thiserror.workspace = true
smallvec.workspace = true

# Using workspace dependency for stwo
stwo-prover.workspace = true
