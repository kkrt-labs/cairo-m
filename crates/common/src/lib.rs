pub mod instruction;
pub mod opcode;
pub mod program;

pub use instruction::Instruction;
pub use opcode::Opcode;
pub use program::{Program, ProgramMetadata};
