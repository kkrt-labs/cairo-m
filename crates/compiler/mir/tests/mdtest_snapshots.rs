//! MIR snapshot tests for mdtest cases.
//! This file automatically generates MIR snapshots for all Cairo-M code in the mdtest directory,
//! providing comprehensive coverage of real-world examples through the MIR generation pipeline.

mod common;

use cairo_m_compiler_mir::{generate_mir, PrettyPrint};
use cairo_m_test_utils::{mdtest::MdTestRunner, mdtest_path};
use common::{create_test_crate, TestDatabase};

#[test]
fn test_mdtest_mir_snapshots() {
    use insta::{assert_snapshot, glob, with_settings};

    // Use glob! to discover all markdown files in mdtest directory
    glob!(mdtest_path().to_str().unwrap(), "**/*.md", |path| {
        let db = TestDatabase::default();

        let runner = MdTestRunner::new("MIR", |source, name| {
            let crate_id = create_test_crate(&db, source, name, "mdtest");

            match generate_mir(&db, crate_id) {
                Ok(module) => Ok(module.pretty_print(0)),
                Err(diagnostics) => Err(format!(
                    "MIR generation failed with diagnostics:\n{:#?}",
                    diagnostics
                )),
            }
        });

        let snapshots = runner.run_file(path);

        for snapshot in snapshots {
            with_settings!({
                description => format!("MIR snapshot for mdtest: {}", snapshot.name).as_str(),
                omit_expression => true,
                snapshot_suffix => snapshot.suffix,
                prepend_module_to_snapshot => false,
            }, {
                assert_snapshot!(snapshot.content);
            });
        }
    });
}
