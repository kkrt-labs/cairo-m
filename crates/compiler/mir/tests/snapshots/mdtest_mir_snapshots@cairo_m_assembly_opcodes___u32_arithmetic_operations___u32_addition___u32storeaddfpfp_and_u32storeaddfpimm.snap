---
source: crates/compiler/mir/tests/mdtest_snapshots.rs
description: "MIR snapshot for mdtest: Cairo-M Assembly Opcodes - U32 Arithmetic Operations - U32 Addition - U32StoreAddFpFp and U32StoreAddFpImm"
input_file: mdtest/06-internals/01-opcodes.md
---
Source:
fn test_u32_addition() -> u32 {
    let a: u32 = 100;
    let b: u32 = 50;

    // U32StoreAddFpFp: u32([fp + dst_off], [fp + dst_off + 1]) = u32([fp + src0_off], [fp + src0_off + 1]) + u32([fp + src1_off], [fp + src1_off + 1])
    let sum1: u32 = a + b;

    // U32StoreAddFpImm: u32([fp + dst_off], [fp + dst_off + 1]) = u32([fp + src_off], [fp + src_off + 1]) + u32(imm_lo, imm_hi)
    let sum2: u32 = a + 25;

    return sum1 + sum2;
}
============================================================
Generated MIR:
module {
  // Function 0
  fn test_u32_addition {
    entry: 0

    0:
      %0 = 100 (u32)
      %1 = 50 (u32)
      %2 = %0 U32Add %1
      %3 = %0 U32Add 25
      %4 = %2 U32Add %3
      return %4

  }

}
